File: nodecore-spv/src/main/java/veriblock/model/TransactionPool.java
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package veriblock.model;

import org.veriblock.sdk.models.Sha256Hash;

import java.lang.ref.Reference;
import java.lang.ref.ReferenceQueue;
import java.lang.ref.WeakReference;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

import static com.google.common.base.Preconditions.checkNotNull;

public class TransactionPool {
    private static class WeakPoolReference extends WeakReference<TransactionMeta> {
        public Sha256Hash hash;
        public WeakPoolReference(TransactionMeta meta, ReferenceQueue<TransactionMeta> queue) {
            super(meta, queue);
            hash = meta.getTxId();
        }
    }

    private final ReentrantLock lock = new ReentrantLock(true);
    private final ReferenceQueue<TransactionMeta> referenceQueue = new ReferenceQueue<>();
    private final LinkedHashMap<Sha256Hash, WeakPoolReference> pool = new LinkedHashMap<Sha256Hash, WeakPoolReference>() {
        @Override
        protected boolean removeEldestEntry(Map.Entry<Sha256Hash, WeakPoolReference> eldest) {
            return size() > 1000;
        }
    };

    public int record(Sha256Hash txId, String peerAddress) {
        lock.lock();
        try {
            purge();
            TransactionMeta tx = getOrCreate(txId);
            boolean wasNew = tx.recordBroadcast(peerAddress);
            if (wasNew) {
                // TODO: Add listener for a newly seen tx

                return tx.getBroadcastPeerCount();
            } else {
                return 0;
            }
        } finally {
            lock.unlock();
        }
    }

    public TransactionMeta getOrCreate(Sha256Hash txId) {
        checkNotNull(txId);

        lock.lock();
        try {
            WeakPoolReference reference = pool.get(txId);
            if (reference != null) {
                TransactionMeta transactionMeta = reference.get();
                if (transactionMeta != null) {
                    return transactionMeta;
                }
            }

            TransactionMeta tx = new TransactionMeta(txId);
            pool.put(txId, new WeakPoolReference(tx, referenceQueue));
            return tx;
        } finally {
            lock.unlock();
        }
    }

    public void insert(TransactionMeta meta) {
        checkNotNull(meta);

        lock.lock();
        try {
            pool.put(meta.getTxId(), new WeakPoolReference(meta, referenceQueue));
        } finally {
            lock.unlock();
        }
    }

    private void purge() {
        lock.lock();
        try {
            Reference<? extends TransactionMeta> reference;
            while ((reference = referenceQueue.poll()) != null) {
                WeakPoolReference txReference = (WeakPoolReference)reference;
                pool.remove(txReference.hash);
            }
        } finally {
            lock.unlock();
        }
    }
}

File: nodecore-spv/src/main/java/veriblock/model/TransactionPool.kt
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package veriblock.model

import com.google.common.base.Preconditions
import org.veriblock.sdk.models.Sha256Hash
import java.lang.ref.Reference
import java.lang.ref.ReferenceQueue
import java.lang.ref.WeakReference
import java.util.LinkedHashMap
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class TransactionPool {
    private class WeakPoolReference(
        meta: TransactionMeta,
        queue: ReferenceQueue<TransactionMeta>?
    ) : WeakReference<TransactionMeta>(meta, queue) {
        var hash: Sha256Hash? = meta.txId
    }

    private val lock = ReentrantLock(true)
    private val referenceQueue = ReferenceQueue<TransactionMeta>()
    private val pool: LinkedHashMap<Sha256Hash, WeakPoolReference> = object : LinkedHashMap<Sha256Hash, WeakPoolReference>() {
        override fun removeEldestEntry(eldest: Map.Entry<Sha256Hash, WeakPoolReference>): Boolean {
            return size > 1000
        }
    }

    fun record(txId: Sha256Hash, peerAddress: String): Int = lock.withLock {
        purge()
        val tx = getOrCreate(txId)
        val wasNew = tx.recordBroadcast(peerAddress)
        if (wasNew) {
            // TODO: Add listener for a newly seen tx
            tx.broadcastPeerCount
        } else {
            0
        }
    }

    fun getOrCreate(txId: Sha256Hash): TransactionMeta = lock.withLock {
        val reference = pool[txId]
        if (reference != null) {
            val transactionMeta = reference.get()
            if (transactionMeta != null) {
                return transactionMeta
            }
        }
        val tx = TransactionMeta(txId)
        pool[txId] = WeakPoolReference(tx, referenceQueue)
        tx
    }

    fun insert(meta: TransactionMeta) = lock.withLock {
        pool[meta.txId] = WeakPoolReference(meta, referenceQueue)
    }

    private fun purge() = lock.withLock {
        var reference: Reference<out TransactionMeta>? = referenceQueue.poll()
        while (reference != null) {
            val txReference = reference as WeakPoolReference
            pool.remove(txReference.hash)
            reference = referenceQueue.poll()
        }
    }
}
